src/csharp/Grpc.HealthCheck/HealthServiceImpl.cs

```diff
@@ -86,17 +123,147 @@ public void ClearAll()
         /// <returns>The asynchronous response.</returns>
         public override Task<HealthCheckResponse> Check(HealthCheckRequest request, ServerCallContext context)
         {
-            lock (myLock)
+            HealthCheckResponse response = GetHealthCheckResponse(request.Service, throwOnNotFound: true);
+
+            return Task.FromResult(response);
+        }
+
+#if GRPC_SUPPORT_WATCH
+        /// <summary>
+        /// Performs a watch for the serving status of the requested service.
+        /// The server will immediately send back a message indicating the current
+        /// serving status.  It will then subsequently send a new message whenever
+        /// the service's serving status changes.
+        ///
+        /// If the requested service is unknown when the call is received, the
+        /// server will send a message setting the serving status to
+        /// SERVICE_UNKNOWN but will *not* terminate the call.  If at some
+        /// future point, the serving status of the service becomes known, the
+        /// server will send a new message with the service's serving status.
+        ///
+        /// If the call terminates with status UNIMPLEMENTED, then clients
+        /// should assume this method is not supported and should not retry the
+        /// call.  If the call terminates with any other status (including OK),
+        /// clients should retry the call with appropriate exponential backoff.
+        /// </summary>
+        /// <param name="request">The request received from the client.</param>
+        /// <param name="responseStream">Used for sending responses back to the client.</param>
+        /// <param name="context">The context of the server-side call handler being invoked.</param>
+        /// <returns>A task indicating completion of the handler.</returns>
+        public override async Task Watch(HealthCheckRequest request, IServerStreamWriter<HealthCheckResponse> responseStream, ServerCallContext context)
+        {
+            string service = request.Service;
+
+            HealthCheckResponse response = GetHealthCheckResponse(service, throwOnNotFound: false);
+            await responseStream.WriteAsync(response);
+
+            // Channel is used to to marshall multiple callers updating status into a single queue.
+            // This is required because IServerStreamWriter is not thread safe.
+            // The channel will buffer up to XXX messages, after which it will drop the oldest messages.
+            Channel<HealthCheckResponse> channel = Channel.CreateBounded<HealthCheckResponse>(new BoundedChannelOptions(capacity: 5) {
```

Comment:
add a comment what's the semantics of "dropping the oldest messages" in terms of the health check protocol.
What happens if the service status is changing more quickly than the Watch method can send them out?
It looks like the intermediate states will be forgotten and it will try to send out the most recent statuses instead?

Reply:
nit: declare a const  with the default capacity and add a comment on what the semantics is.